File: src/lib/api/aiProvider.ts
import { devApiFetch, hasTauriRuntime, invoke } from "./client";
import type {
  AiProvider,
  AiProviderAskInput,
  AiProviderAskResult,
  AiProviderModelListInput,
  AiProviderModelListResult,
  AiProviderRoute,
  AiProviderTestResult,
  UpsertAiProviderInput,
  UpsertAiProviderRouteInput,
} from "@/types";

export const aiProviderApi = {
  list: () =>
    hasTauriRuntime()
      ? invoke<AiProvider[]>("list_ai_providers")
      : devApiFetch<AiProvider[]>("/ai-providers"),
  upsert: (input: UpsertAiProviderInput) =>
    hasTauriRuntime()
      ? invoke<AiProvider>("upsert_ai_provider", { input })
      : devApiFetch<AiProvider>("/ai-providers", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  delete: (key: string) =>
    hasTauriRuntime()
      ? invoke<void>("delete_ai_provider", { key })
      : devApiFetch<void>(`/ai-providers/${encodeURIComponent(key)}`, {
          method: "DELETE",
        }),
  listRoutes: () =>
    hasTauriRuntime()
      ? invoke<AiProviderRoute[]>("list_ai_provider_routes")
      : devApiFetch<AiProviderRoute[]>("/ai-providers/routes"),
  upsertRoute: (input: UpsertAiProviderRouteInput) =>
    hasTauriRuntime()
      ? invoke<AiProviderRoute>("upsert_ai_provider_route", { input })
      : devApiFetch<AiProviderRoute>("/ai-providers/routes", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  test: (key: string) =>
    hasTauriRuntime()
      ? invoke<AiProviderTestResult>("test_ai_provider", { key })
      : devApiFetch<AiProviderTestResult>(
          `/ai-providers/${encodeURIComponent(key)}/test`,
          { method: "POST" },
        ),
  listModels: (input: AiProviderModelListInput) =>
    hasTauriRuntime()
      ? invoke<AiProviderModelListResult>("list_ai_provider_models", { input })
      : devApiFetch<AiProviderModelListResult>("/ai-providers/models", {
          method: "POST",
          body: JSON.stringify(input),
        }),
  ask: (input: AiProviderAskInput) =>
    hasTauriRuntime()
      ? invoke<AiProviderAskResult>("ask_ai_provider", { input })
      : devApiFetch<AiProviderAskResult>("/ai-providers/ask", {
          method: "POST",
          body: JSON.stringify(input),
        }),
};

File: src-tauri/src/commands/ai_provider.rs
use crate::error::CommandError;
use crate::models::{
    AiProvider, AiProviderAskInput, AiProviderAskResult, AiProviderModelListInput,
    AiProviderModelListResult, AiProviderRoute, AiProviderTestResult, UpsertAiProviderInput,
    UpsertAiProviderRouteInput,
};
use crate::services::ai_provider::AiProviderService;
use crate::state::AppState;

#[tauri::command]
pub fn list_ai_providers(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<AiProvider>, CommandError> {
    AiProviderService::list(&state.db).map_err(|e| e.into())
}

#[tauri::command]
pub fn upsert_ai_provider(
    state: tauri::State<'_, AppState>,
    input: UpsertAiProviderInput,
) -> Result<AiProvider, CommandError> {
    AiProviderService::upsert(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub fn delete_ai_provider(
    state: tauri::State<'_, AppState>,
    key: String,
) -> Result<(), CommandError> {
    AiProviderService::delete(&state.db, &key).map_err(|e| e.into())
}

#[tauri::command]
pub fn list_ai_provider_routes(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<AiProviderRoute>, CommandError> {
    AiProviderService::list_routes(&state.db).map_err(|e| e.into())
}

#[tauri::command]
pub fn upsert_ai_provider_route(
    state: tauri::State<'_, AppState>,
    input: UpsertAiProviderRouteInput,
) -> Result<AiProviderRoute, CommandError> {
    AiProviderService::upsert_route(&state.db, input).map_err(|e| e.into())
}

#[tauri::command]
pub async fn test_ai_provider(
    state: tauri::State<'_, AppState>,
    key: String,
) -> Result<AiProviderTestResult, CommandError> {
    AiProviderService::test(&state.db, &key)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn list_ai_provider_models(
    state: tauri::State<'_, AppState>,
    input: AiProviderModelListInput,
) -> Result<AiProviderModelListResult, CommandError> {
    AiProviderService::list_models(&state.db, input)
        .await
        .map_err(|e| e.into())
}

#[tauri::command]
pub async fn ask_ai_provider(
    state: tauri::State<'_, AppState>,
    input: AiProviderAskInput,
) -> Result<AiProviderAskResult, CommandError> {
    AiProviderService::ask(&state.db, input)
        .await
        .map_err(|e| e.into())
}
